### PR TITLE
Allow deep update checking for collection update server

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/25766

### Summary of Changes

This here now implements the option to let joomla do a deep update checking over collection update server results.

As pointed out here https://github.com/joomla/joomla-cms/issues/25766#issuecomment-520575725 the collection update server does not support any limiting options like php version or minimum stabilty checks etc.
The reason is that for the initial check only the data of the collection file is used and respected. Only at the actual install time Joomla checks the extension.xml wich than might point to an different version or to no version at all because of the limiting parameters in the extension.xml

This here now allow the extension dev to configure Joomla to do this deep update checking by enable the deep update checking for the collection updateserver.

#### First Test

- install joomla
- install the dummy plugin: https://www.jah-tz.de/downloads/download.php?file=plg_system_dummy.zip&folder=core/extension/dummy/
- try to find an update
- you find the update to 1.0.1
- try to install the update
- it fails as the requirement for this plugin is php 8.0.0

#### Second Test

- install joomla
- ~~apply the update SQL from this patch~~
- apply the patch
- install the dummy plugin: https://www.jah-tz.de/downloads/download.php?file=plg_system_dummy.zip&folder=core/extension/dummy/
- try to find an update
- joomla tells you that it found an update but it requires PHP 8.0.0

### Expected result

image

### Actual result

Joomla shows the update but can't install it.

image

### Documentation Changes Required

- [ ] The new option for collection update servers need to be documented.

### Additional Questions

As of now I have implemented this new option to be a update site wide decision. We could it also change the implementation to just check in the XML for this new option so no need for database changes etc and you just define the parameter in the collection.xml itself and joomla does the deep update checking, This way the site owner could not overwrite this setting. Thougths? Optinions?

### CODE REVIEW

**Comment by ghost** [comment]
@alex7r please test.

**Comment by brianteeman** [comment]
is this different to the new pre-update check feature

**Comment by zero-24** [comment]
> is this different to the new pre-update check feature

Yes as this is primary intended to be used by extension update servers.

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 1]
```diff
@@ -4,6 +4,8 @@
 ; Note : All ini files need to be saved as UTF-8
 
 COM_JOOMLAUPDATE_CHECKED_UPDATES="Checked for updates."
+COM_JOOMLAUPDATE_CONFIG_CHECKDETAILSURL_LABEL="Enable Deep Update checking"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: I think you are going to have to come up with a better name than "Deep Update" as its pretty meaningless. And no I can't suggest something as I dont understand it either

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: Well what we actually do is that we check the details URL and make sure the update the collection pointed us to is available to our system. We could also name it `Extented Update checks` or something similar. In the end we just make sure that the update can be installed on the system by checking the provided extension.xml too.

> And no I can't suggest something as I dont understand it either

Hmm can you tell me how far you understand what I did? So I can help with the missing parts.

**Comment by mbabker** [comment]
1) Why do you need a new database field?
2) Why are you using camelCase on the field when most of the database schema uses snake_case?
3) The current naming implies it is storing a URL when in reality it's some boolean toggle
4) Why does this need to be a toggle?
5) Until someone writes some proper tests, can we stop adding features to or making changes to the `Joomla\CMS\Updater` classes because right now it is beyond a pain in the ass to test and validate these things actually work and this critical library has absolutely zero coverage (a task for the testing team if they decide to actually write tests and stop working on architecture or devoting time to things that have nothing to do with testing).

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini (hunk 1 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: I am still confused about the difference with this and the pre-update check in 3.10

Why does it even need an option? Would you ever want/need to disable it?

**Comment by zero-24** [comment]
> Why do you need a new database field?

As said above it is not an requirement it was just the first idea, can be an option in the XML too.

> Why are you using camelCase on the field when most of the database schema uses snake_case?

As it is going to be a option in the XML on your suggestion, this issue can be migrated by the new name.

> The current naming implies it is storing a URL when in reality it's some boolean toggle

I can come up with a different name.

> Why does this need to be a toggle?

Well for the most update servers this is not relevant so the idea was to make this optional.

> Until someone writes some proper tests, can we stop adding features to or making changes to the Joomla\CMS\Updater

Ok, sadly I can't help here for now. :(

**Comment by zero-24** [comment]
ok just double checked the code. With the current architektur of the updater I don't see a way to overcome some sort of database table change as any field not in the database table is dropped [here](https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Updater/Adapter/CollectionAdapter.php#L185)

So we could add an additional field to the #__updates or the #__update_sites table but without bigger changes to the updater i don't see a way to overcome that thing.

The other way around would be to implement the checks just before we add the update to the database but I'm not sure whether this actually works.

**Comment by zero-24** [comment]
Ok have just addressed most of the questions raised here.

I have now removed the sql changes and now use a dedicated setting in the collection.xml (`forcedeepextensionchecking="1"`). So nothing else is required and it can be set per line in the collection.xml. For sure this is still an optional thing an has to be configured by the extension dev, so no need anymore for an switch in the UI.

The only open point is the Unit test I'm happy to build tests but i have no idea how to do it, so i'm happy for some help me on that. Technical wise this PR here is already working and for sure gets documented by me when the changes here are merged into the core.

I'm happy for feedback on the new / improved implementation 👍

File: libraries/src/Updater/Updater.php [hunk 2]
```diff
@@ -321,6 +321,45 @@ private function getUpdateObjectsForSite($updateSite, $minimum_stability = self:
...
 				{
+					// Check with the extension adapter that the update here is actually an update we can install
+					if ($updateSite['type'] === 'collection'
+						&& is_array($current_update->data)
+						&& isset($current_update->data['forcedeepextensionchecking'])
+						&& (bool) $current_update->data['forcedeepextensionchecking'] === true)
+					{
+						/*
+						 * Using the forcedeepextensionchecking parameter this collection update server request
+						 * use to check the update with the extension adapter too. Let's do this
```
Comment (brianteeman) [inline libraries/src/Updater/Updater.php]:
```suggestion
						 * is used to check the update with the extension adapter. Let's do this
```

Comment (zero-24) [reply libraries/src/Updater/Updater.php]: Would you agree with `Using the forcedeepextensionchecking parameter this collection update server request us to check the update with the extension adapter too. Let's do this`

It was intended to be `us` vs `use` ;)

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on b41f8367d202e236a8943715485a86a2544676eb<br /><br />tested: except the point: apply the update SQL <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25928">issues.joomla.org/tracker/joomla-cms/25928</a>.</sub>

**Comment by zero-24** [comment]
Yes and thanks for your test! @ChristineWk 

>  except the point: apply the update SQL

This is correct as changes to the sql is no longern needed :)

**Comment by zero-24** [comment]
@alex7r It would be great to get your feedback here. 👍

**Comment by zero-24** [comment]
@wilsonge @HLeithner @mbabker  As the original reporter did not come back do you guy's think this is a feature we should still implement or should this PR be closed and the feature dropped?

When the feature is not included we would still have the issue with the collection.xml not checking the deeper requirements of an update that could also hit us at some time in the future with the core update as we use the collection updates for our updating. And this feature is also opt-in for the extension dev so we are also b/c for someone relying on the issue we are fixing here.

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on bc1d809dc8e55e75f290538c9c9a90fad5bd09c2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25928">issues.joomla.org/tracker/joomla-cms/25928</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25928">issues.joomla.org/tracker/joomla-cms/25928</a>.</sub>

**Comment by HLeithner** [comment]
I will postpone to the next release and further discussion.

**Comment by zero-24** [comment]
Ok so better close here @HLeithner ?

**Comment by zero-24** [comment]
@HLeithner any update on this one?

**Comment by zero-24** [comment]
@HLeithner @wilsonge what is you opinion on this is it worth to keep it open for longer or should it just be closed and documented as expected behavior?

**Comment by wilsonge** [comment]
This is one of the PRs i have open in a tab to review I mentioned it maintainers chat. Honestly I didn’t understand the term deep update checking either so give me a week to sit down with the code and fully understand what the change is here please and I’ll get back to you

**Comment by zero-24** [comment]
Great, just let me know any question that comes up. In a nutshell with deep update checking we apply the same rules and checks for example the php minimum or stability to updates offered via an collection update server. As of today the update just silently fails as joomla cant get the download url at the actual update step. ;-)

**Review by wilsonge — APPROVED** [review]
Looks fine to me - although unsure we should actually put this behind a flag and not just make this default behaviour

**Comment by zero-24** [comment]
As discussed the latest changes makes this a client site decision enabled by default. thanks @wilsonge

**Review by wilsonge — APPROVED** [review]

**Comment by adj9** [comment]
With PHP 7.3.19 there is no plugin update, PHP 8 cannot be enabled yet.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25928">issues.joomla.org/tracker/joomla-cms/25928</a>.</sub>

**Comment by zero-24** [comment]
@adj9 Thanks for your test so you get the message as posted in the inital post right?

It is intended that ("without php 8.0.0") that message comes up.

When it was that case please log your successful test on https://issues.joomla.org

**Comment by zero-24** [comment]
Ah just saw that update server did not server to 3.10 installs as this here was inital build against 3.9 that should be fixed now.

**Comment by zero-24** [comment]
Seems there is still no interest in getting this or similiar PR merged. No need to keep that updated. Closing.

**Pull request closed without merging**